File: app/twitch/ingress/test/nats_credentials_test.exs
# Copyright (c) 2026 Adam Ousmer. All rights reserved.
# Proprietary. No license granted. See LICENSE.md.

defmodule Ingress.NatsCredentialsTest do
  use ExUnit.Case, async: false

  @runtime "config/runtime.exs"
  @keys [
    "NATS_USER",
    "NATS_PASSWORD",
    "NATS_JWT",
    "NATS_NKEY_SEED",
    "NATS_RPC_JWT",
    "NATS_RPC_NKEY_SEED"
  ]

  setup do
    System.put_env("NATS_JWT", "bus-jwt")
    System.put_env("NATS_NKEY_SEED", "bus-seed")
    on_exit(fn -> Enum.each(@keys, &System.delete_env/1) end)
    :ok
  end

  test "both planes send the JWT and seed, never a leftover password" do
    System.put_env("NATS_USER", "bus-user")
    System.put_env("NATS_PASSWORD", "bus-pass")

    config = read_runtime()

    for plane <- [:nats, :nats_bus] do
      assert [%{jwt: "bus-jwt", nkey_seed: "bus-seed"} = server] = config[:ingress][plane]
      refute Map.has_key?(server, :username)
      refute Map.has_key?(server, :password)
    end
  end

  test "an RPC-specific JWT pair wins over the bus pair on the RPC plane only" do
    System.put_env("NATS_RPC_JWT", "rpc-jwt")
    System.put_env("NATS_RPC_NKEY_SEED", "rpc-seed")

    config = read_runtime()

    assert [%{jwt: "rpc-jwt", nkey_seed: "rpc-seed"}] = config[:ingress][:nats]
    assert [%{jwt: "bus-jwt", nkey_seed: "bus-seed"}] = config[:ingress][:nats_bus]
  end

  test "an incomplete pair sends no credentials" do
    System.delete_env("NATS_NKEY_SEED")

    config = read_runtime()

    refute Map.has_key?(List.first(config[:ingress][:nats_bus]), :jwt)
  end

  defp read_runtime do
    {config, _warnings} =
      ExUnit.CaptureIO.with_io(:stderr, fn -> Config.Reader.read!(@runtime, env: :test) end)

    config
  end
end

File: web/kit/lib/server/nats-credentials.test.ts
// Copyright (c) 2026 Adam Ousmer. All rights reserved.
// Proprietary. No license granted. See LICENSE.md.

import { afterEach, beforeEach, describe, expect, mock, test } from 'bun:test';

mock.module('newrelic', () => ({
  default: {
    startSegment: (_name: string, _record: boolean, run: () => unknown) => run(),
    recordMetric: () => {}
  }
}));
const { credentialAuth, options } = await import('./nats');

// Throwaway user nkey seed generated for this test only; not a live credential.
const SEED = 'SUAIVAGIKFP625PEBOON4MYTSXVJKY3TIGDMTGNYW47VW45MNW4EWO32KQ';

const CREDENTIAL_ENV_KEYS = [
  'NATS_USER',
  'NATS_PASSWORD',
  'NATS_JWT',
  'NATS_NKEY_SEED',
  'NATS_RPC_JWT',
  'NATS_RPC_NKEY_SEED'
];
const saved: Record<string, string | undefined> = {};

beforeEach(() => {
  for (const key of CREDENTIAL_ENV_KEYS) {
    saved[key] = process.env[key];
    delete process.env[key];
  }
});

afterEach(() => {
  for (const key of CREDENTIAL_ENV_KEYS) {
    if (saved[key] === undefined) delete process.env[key];
    else process.env[key] = saved[key];
  }
});

describe('credentialAuth', () => {
  test('a JWT and seed build a single JWT authenticator', () => {
    const auth = credentialAuth('a-jwt', SEED);
    expect(auth.authenticator).toHaveLength(1);
    const creds = auth.authenticator![0]('nonce') as { jwt: string; sig: string };
    expect(creds.jwt).toBe('a-jwt');
    expect(creds.sig.length).toBeGreaterThan(0);
  });

  test('either var alone yields an empty options patch', () => {
    expect(credentialAuth('a-jwt', undefined)).toEqual({});
    expect(credentialAuth(undefined, SEED)).toEqual({});
  });
});

describe('options() credential wiring', () => {
  test('leftover password vars are never sent', () => {
    process.env.NATS_USER = 'u';
    process.env.NATS_PASSWORD = 'p';
    process.env.NATS_JWT = 'bus-jwt';
    process.env.NATS_NKEY_SEED = SEED;

    const opts = options('bus');
    expect(opts.user).toBeUndefined();
    expect(opts.pass).toBeUndefined();
    expect(opts.authenticator).toHaveLength(1);
  });

  test('RPC falls back to the bus JWT vars', () => {
    process.env.NATS_JWT = 'bus-jwt';
    process.env.NATS_NKEY_SEED = SEED;

    const opts = options('rpc');
    expect(opts.authenticator).toHaveLength(1);
    const creds = opts.authenticator![0]('nonce') as { jwt: string };
    expect(creds.jwt).toBe('bus-jwt');
  });

  test('RPC-specific JWT vars win over the bus fallback', () => {
    process.env.NATS_JWT = 'bus-jwt';
    process.env.NATS_NKEY_SEED = SEED;
    process.env.NATS_RPC_JWT = 'rpc-jwt';
    process.env.NATS_RPC_NKEY_SEED = SEED;

    const creds = options('rpc').authenticator![0]('nonce') as { jwt: string };
    expect(creds.jwt).toBe('rpc-jwt');
  });

  test('the bus role never reads the RPC-specific JWT vars', () => {
    process.env.NATS_RPC_JWT = 'rpc-jwt';
    process.env.NATS_RPC_NKEY_SEED = SEED;

    expect(options('bus').authenticator).toBeUndefined();
  });
});
